### PR TITLE
DS-3905 Lookup DSpace Object by UUID

### PR DESCRIPTION
**Pull request by tomdesair**
The Angular UI needs a generic way to lookup any DSpace Object based on a UUID. This is also necessary to support "Find by item ID".

We've extended the original implementation of DS-3905 to also support UUID identifiers. We've added integration tests to cover these scenario's.

We've removed the `org.dspace.identifier.Identifier` interface because it was not actually used. And there was no way to make `java.util.UUID` extend that interface.

### CODE REVIEW

File: dspace-api/src/main/java/org/dspace/identifier/UUIDIdentifierProvider.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import org.dspace.content.DSpaceObject;
+import org.dspace.content.factory.ContentServiceFactory;
+import org.dspace.content.service.DSpaceObjectService;
+import org.dspace.core.Context;
+import org.dspace.util.UUIDUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class UUIDIdentifierProvider extends IdentifierProvider {
```
Comment (tdonohue) [inline dspace-api/src/main/java/org/dspace/identifier/UUIDIdentifierProvider.java]: Sorry, I'm going to sound like a "broken record".  Javadocs please ;)

**Review by tdonohue — APPROVED** [review]
👍 Code looks good overall.  Just my usual request for Javadocs for new public classes / methods (see inline below).

**Review by terrywbrady — APPROVED** [review]
The code looks good to me.  Tested locally and verified uuid retrieval in rest.

**Comment by abollini** [comment]
I'm not sure that is a good idea to implement an UUID Identifier provider, it seems to me a bit abused concept just to implement the lookup. This is confirmed by the many "return null" implementation in the UUIDIdentifierProvider.
We end dropping the Identifier mark interface, that beside haven't a functional value allow us to easily recognize which are the supported Identifier types. Saying that the uuid is an identifier lead in some way to the suggestion to use it for permanent linking that is something that we don't want (right?).

What about implementing a simple /core/dsos/{uuid} readonly endpoint or another custom endpoint parallel to the pid one?
I'm not strongly against that, I like to hear other opinion.

**Comment by mwoodiupui** [comment]
We certainly should not expose UUID as an identifier with meaning outside of a single DSpace instance.  If we do, we will need to invent another internal-only identifier.

If UUID must be usable at all above the business logic layer, it should be a simple one-way operation:  if you know a UUID then you can look up the object identified, but you cannot enumerate UUIDs.  Andrea's proposal seems to meet that requirement.

**Comment by terrywbrady** [comment]
In DSpace 6, the UUID is used as the key for Bulk Metadata Updates.  Unless there is another option for updating metadata, this key needs to be exposed.

**Comment by tdonohue** [comment]
Just to clarify, I don't believe this implementation allows for enumerating UUIDs (not seeing it in the code, but maybe I overlooked it).  It is a simply a lookup of an object using its UUID.  That is definitely a feature we'll need/want in the REST API layer (as UUIDs are the only guaranteed object identifier, especially for things like WorkspaceItems and WorkflowItems)

That said, I see @abollini's point that a `UUIDIdentifierProvider` is a somewhat "awkward" way of implementing this lookup.  That IdentifierProvider doesn't allow you to register/mint a new UUID (both methods return null) as that code is elsewhere in DSpace...and all it does is allow for a `resolve` and `lookup`.  

While the UUID is technically a type of identifier, it does seem a different treatment from Handles or DOIs, in that UUIDs are very much *required* and integral to DSpace.  We cannot even *create* objects without a UUID (as it's required at the database layer).  So, it does seem a bit odd to align UUID with other external, non-required Identifiers.

So, I think I'm starting to agree with @abollini that we may want to rethink this implementation, and simply provide another way to lookup an object via its UUID.

**Comment by tomdesair** [comment]
While I agree that the `UUIDIdentifierProvider` implementation looks strange because of the empty methods, it has the advantages that:

- It provides a generic "lookup by UUID" function within the `dspace-api` module so that other `dspace-api` classes and DSpace modules can reuse this.
- There is only one endpoint that resolves (any) identifier to the corresponding object. So there is no confusion on which endpoint to use if you do not know the DSO type.

**Review by abollini — CHANGES_REQUESTED** [review]
@tomdesair the benefits that you point out can be IMHO accomplished without introduce a UUIDIdentiferProvider
- the generic "lookup by UUID" function is already here is just the dspaceobject service that the UUIDIdenfierProvider will call
https://github.com/DSpace/DSpace/pull/2125/files#diff-12eb4667ea870a15225b781df526c0f4R56

- as said I'm not against an endpoint that resolve any uuid to the corresponding DSpaceObject without previous knowledge about the DSO type. I'm just arguing that it could be "cleaner" to have it separated from the /pid/ endpoint

I'm adding this comment as a "formal review" requiring changes to proper flag this PR as "still under discussion"  but I will be happy to have that merged if other are convinced by this direction

**Comment by tomdesair** [comment]
@abollini Please note that just above the line you mentioned there is a for-loop:

```
for (DSpaceObjectService dSpaceObjectService : contentServiceFactory.getDSpaceObjectServices()) {
```

We did not find a generic `DSpaceObjectService` implementation as the dspace-api module does not seem to provide a generic "lookup by UUID". So it is incorrect to state that the `UUIDIdenfierProvider` just calls _the_ `DSpaceObjectService`. In fact it will try *every* possible `DSpaceObjectService` implementation.

So the `UUIDIdentifierProvider` can still be a useful addition for the dspace-api module as it does provide a new generic "lookup by UUID" functionality.

**Comment by tdonohue** [comment]
As @tomdesair noted, this generic lookup doesn't exist elsewhere, as we don't have a generic `DSpaceObjectServiceImpl.find()` method.  If we did have that method, it'd require the same sort of code as is in the `UUIDIdenfierProvider.resolve()` method in this PR: https://github.com/DSpace/DSpace/pull/2125/files#diff-12eb4667ea870a15225b781df526c0f4R48

That said, this method logic looks odd to me at a second glance.  To lookup something generically by just its UUID, we have to perform *up to* 5 separate queries (against Bitstream, Bundle, Item, Collection, Community).  That implies that even if we made this endpoint available, we'd want to warn against using it heavily -- it's MUCH better to look something up by UUID via the existing DSO endpoints (`/api/core/[dso]/<:uuid>`)

Taking a step back here, I'm not sure I completely understand *why* we need a generic UUID lookup.  The original use case described here by @tomdesair  (in the description) was for "Find by Item ID" (which  can just use `/api/core/items/<:uuid>` as we already know it's an Item).

Do we have a clear use case where the Angular UI actually needs to look something up by UUID *and doesn't know the DSO type*?

**Comment by artlowel** [comment]
>Do we have a clear use case where the Angular UI actually needs to look something up by UUID and doesn't know the DSO type?

Yes, when you arrive at the search page with a scope filter already active,
all we know about the scope from the URL is its UUID. We can't know whether it's a Community or a Collection. I asked @tomdesair for this PR for that reason. It is used [here](https://github.com/DSpace/dspace-angular/pull/280/files#diff-4ef3e816a518eb897da43f7f538cf230R270) in the search features PR

**Comment by tdonohue** [comment]
@artlowel : thanks for the note on where/how this is used in Angular UI.  I see the point of it now, and it's definitely necessary.

👍 I think this implementation looks fine to me.  We can always change the endpoint in the future, if we have a better reason to do so.  

However, that said, before merger I'd *strongly recommend we provide warnings* (in both JavaDocs and REST Contract) against usage of this new `/api/pid/find?id=<:uuid>` endpoint (except where absolutely necessary).  Based on the code in this PR, it seems like performance _may be significantly worse_ than, for example using `/api/core/items/<:uuid>` (for Items), `/api/core/collections/<:uuid>` (for Collections), etc.

I'd just hate to see Angular UI developers (or others building tools against the REST API) get lazy and begin using `/api/pid/find?id=<:uuid>` in situations where the object type is known.  Doing so will definitely have a performance effect on large sites.  We want to only use this endpoint sparingly or fix the performance issue in the backend database layer.

@paulo-graca : You are correct, it's too soon to talk about this in relationship to entities. I think that will depend on the final entities implementation. So, I don't want to speculate until we know more.  But, it very well may be that entities efforts may provide an excuse to fix the backend performance issues here.

**Comment by Raf-atmire** [comment]
@tdonohue I've added JavaDocs to the UUIDIdentifierProvider class and I've included a disclaimer about the performance issues that using this class abundantly may incur. If there is a more visually apparent way to display this warning, I'll be happy to adjust the documentation accordingly.

**Comment by tdonohue** [comment]
@abollini : Any final thoughts here (with regards to your requested changes)? 

I'm inclined to merge this as-is next week (as it is currently blocking the review/merger of DSpace/dspace-angular#280), unless you still have strong objections.  I feel we could always clean this up more in a future PR, and I'd hate to continue to block other code that depends on a solution/decision here.

**Comment by abollini** [comment]
To clarify my position
- I'm in favor to introduce a findByUUID endpoint in the REST API as *fallback* in scenario where we don't know the dso type in advance
- I'm still not convinced that this is a PID (the endpoint is /pid/find). I will be more happy to have that on a separate endpoint. This will also remove the need to have the implementation to extends IdentifierProvider (that sound really odd). It should be easy to move the current implementation of the UUIDIdentifierProvider.resolve method https://github.com/DSpace/DSpace/pull/2125/files#diff-12eb4667ea870a15225b781df526c0f4R81 in a dedicated utils class and have a new REST controller to simply use it. Of course the utils class will be available to other dspace-api code as well;
- in regards to the angular PR https://github.com/DSpace/DSpace/pull/2125#issuecomment-410616793 @artlowel should be possible to keep the dso type in the URL parameter just to avoid the performance issue? this is aligned with the choice reported by @tdonohue  to still have the dso type in some table to speedup the lookup process

That said my "request changes" review is NOT a veto

**Comment by tdonohue** [comment]
We discussed this in today's DSpace 7 Meeting (see [Notes](https://wiki.duraspace.org/display/DSPACE/2018-08-16+DSpace+7+Working+Group+Meeting)).

@abollini and @ppmdo were going to look at achieving this endpoint via a Solr query in the coming days (as this would alleviate any performance concerns listed above). They will report back on progress.  If it is not achievable by or on Monday, we may consider merging this as-is, so as to unblock https://github.com/DSpace/dspace-angular/pull/280.

**Comment by mwoodiupui** [comment]
Is it fair for me to say that, at this point, the performance issues are entirely theoretical?  Because looking up a given DSO type by UUID should be quite cheap, as UUIDs are indexed.  The performance concern is not over an expensive operation, but over doing a cheap operation 2.5 times on average.  That is not very much additional cost in absolute terms.  How much discussion and experimentation and reimplementation is it worth, before we experience an actual cost that we're unwilling to pay?

**Comment by ppmdo** [comment]
@mwoodiupui In the case of disaster recovery, handles are still the best "unique identifiers" for a given object, right?

**Comment by tdonohue** [comment]
@mwoodiupui : I think your concerns about UUIDs are very similar to @abollini.  @abollini said he didn't want this on the `/api/pid/` endpoint because UUIDs are not PIDs (Persistent Identifiers), as UUIDs are not guaranteed to be persistent.

@ppmdo : correct, Handles are guaranteed persistent (even in the AIP Backup and Restore process).  But, UUIDs are definitely not persistent in the AIP Backup & Restore process (in fact there's a ticket about that: https://jira.duraspace.org/browse/DS-3961).

I agree that concerns about performance could be addressed later, if we find that a "quick fix" here isn't possible. That's why we have a timeline of Monday to investigate another solution -- if it's found to not be plausible, then we will go with this one.

**Comment by abollini** [comment]
as by discussion in the yesterday meeting @artlowel @tdonohue I have created the issue https://jira.duraspace.org/browse/DS-3984 
I will appreciated if you (or anyone else) can add examples of code where the dspaceobject is retrieved without having the resource type

**Comment by mwoodiupui** [comment]
Use of UUIDs in batch metadata editing was mentioned above.  There's some discussion of that in Slack today.  I'll repeat here that I think this is an internal use of UUIDs and not an issue.

**Comment by tomdesair** [comment]
As discussed, I'm closing this in favour of https://github.com/DSpace/DSpace/pull/2179

**Pull request closed without merging**